DataStructures/bs_tree.py

```diff
@@ -0,0 +1,181 @@
+                    current.right = new_node
+                    break
+                #if we already have a right node then we update the current node
+                current = current.right
+        return
+
+    def lookup(self, data: Any) -> Node:
+        """
+        Finds a node in the binary search tree by its value and returns the node.
+    
+        Parameters:
+        data: Any - the value to search for in the tree.
+    
+        Returns:
+        Node - the node containing the value, or None if the value is not found.
+        """
+        current = self.root
+        while current:
+            if data == current.data:
+                return current  # Found the value
+            if data < current.data:
+                current = current.left  # Go to the left subtree
+            else:
+                current = current.right  # Go to the right subtree
+        return None  # Value not found
+
+    def delete(self, data: Any, start: Node = None, parent: Node = None) -> None:
+        """
+        Deletes a node with the specified value from the tree.
+
+        Parameters:
+        - data (Any): The value to be deleted.
+        - start (Optional[Node]): The starting point for the search (default: root).
+        - parent (Optional[Node]): The parent node of the current node.
+
+        Raises:
+        - Exception: If the value is not found in the tree.
+        """
+        current = start or self.root
+        while current and current.data != data:
+            parent = current
+            if data < current.data:
+                current = parent.left
+            else:
+                current = parent.right
+        if not current:
+            #if the node is not in tree
+            raise ValueError("Item not in tree")
+
+        #checking the childrens
+        if not current.left and not current.right:
+            #if the node has no childrens
+            return self._remove_node_no_children(current, parent)
+        if current.left and current.right:
+            #if the node has two childrens
+            print (f"Two children node {current.data} successfully deleted")
+            return self._remove_node_two_children(current)
+        #if the node has one child
+        return self._remove_node_one_child(current, parent)
+
+    def _remove_node_no_children(self, current: Node, parent: Node) -> None:
+        """
+        Removes a node with no children (leaf node).
+
+        Parameters:
+        - current (Node): The node to be removed.
+        - parent (Optional[Node]): The parent node of the current node.
+        """
+        if current is self.root:
+            #removing the root if this is the only node in tree
+            self.root = None
+            return self
+        if parent.left == current:
+            parent.left = None
+        else:
+            parent.right = None
+        print(f"Leaf node {current.data} successfully deleted")
+        return self
```

Comment:
In the function definition you are writing that function returns None, but it is actually returns self class object
![image](https://github.com/user-attachments/assets/0c020421-362b-4197-8c31-07757a01d53a)